pkg/sources/s3/s3.go

```diff
@@ -134,76 +175,99 @@ func (s *Source) newClient(region string) (*s3.S3, error) {
-		if len(s.conn.Buckets) == 0 {
-			res, err := client.ListBuckets(&s3.ListBucketsInput{})
+			s.log.Info("Scanning bucket", "bucket", bucket)
+			region, err := s3manager.GetBucketRegionWithClient(context.Background(), client, bucket)
 			if err != nil {
-				return fmt.Errorf("could not list s3 buckets: %w", err)
+				s.log.Error(err, "could not get s3 region for bucket", "bucket", bucket)
+				continue
 			}
-			buckets := res.Buckets
-			for _, bucket := range buckets {
-				bucketsToScan = append(bucketsToScan, *bucket.Name)
+			var regionalClient *s3.S3
+			if region != defaultAWSRegion {
+				regionalClient, err = s.newClient(region)
+				if err != nil {
+					s.log.Error(err, "could not make regional s3 client")
+				}
+			} else {
+				regionalClient = client
 			}
-		} else {
-			bucketsToScan = s.conn.Buckets
-		}
-	case *sourcespb.S3_Unauthenticated:
-		bucketsToScan = s.conn.Buckets
-	default:
-		return errors.Errorf("invalid configuration given for %s source", s.name)
-	}
 
-	objectCount := uint64(0)
-	for i, bucket := range bucketsToScan {
-		if common.IsDone(ctx) {
-			return nil
-		}
+			errorCount := sync.Map{}
 
-		s.SetProgressComplete(i, len(bucketsToScan), fmt.Sprintf("Bucket: %s", bucket), "")
+			err = regionalClient.ListObjectsV2PagesWithContext(
+				ctx, &s3.ListObjectsV2Input{Bucket: &bucket},
+				func(page *s3.ListObjectsV2Output, last bool) bool {
+					s.pageChunker(ctx, regionalClient, chunksChan, bucket, page, &errorCount, i+1, &objectCount)
+					return true
+				})
 
-		s.log.Info("Scanning bucket", "bucket", bucket)
-		region, err := s3manager.GetBucketRegionWithClient(context.Background(), client, bucket)
-		if err != nil {
-			s.log.Error(err, "could not get s3 region for bucket", "bucket", bucket)
-			continue
-		}
-		var regionalClient *s3.S3
-		if region != defaultAWSRegion {
-			regionalClient, err = s.newClient(region)
 			if err != nil {
-				s.log.Error(err, "could not make regional s3 client")
+				return fmt.Errorf(
+					"could not list objects in s3 bucket: bucket %s: %w",
+					bucket,
+					err)
 			}
-		} else {
-			regionalClient = client
-		}
+	}
 
-		errorCount := sync.Map{}
+	s.SetProgressComplete(len(bucketsToScan), len(bucketsToScan), fmt.Sprintf("Completed scanning source %s. %d objects scanned.", s.name, objectCount), "")
 
-		err = regionalClient.ListObjectsV2PagesWithContext(
-			ctx, &s3.ListObjectsV2Input{Bucket: &bucket},
-			func(page *s3.ListObjectsV2Output, last bool) bool {
-				s.pageChunker(ctx, regionalClient, chunksChan, bucket, page, &errorCount, i+1, &objectCount)
-				return true
-			})
+	return nil
+
+	}
 
+	client, err := s.newClient(defaultAWSRegion)
+	if err != nil {
+		return errors.WrapPrefix(err, "could not create s3 client", 0)
+	}
+
+	if err := processBuckets(client); err != nil {
+		return err
+	}
+
+	// If roles are provided, create additional clients for each role and process
+	for _, roleArn := range s.conn.Roles {
```

Comment:
I think we're missing an assignment to the connection's `Roles` attribute in [pkg/engine/s3.go](https://github.com/trufflesecurity/trufflehog/blob/35b332c49860cecd897d4fe5ea39c3eebb08ec81/pkg/engine/s3.go#L21-L23). It should be in the `S3Config` object passed into that function.

https://github.com/trufflesecurity/trufflehog/blob/35b332c49860cecd897d4fe5ea39c3eebb08ec81/pkg/engine/s3.go#L19-L23